Fix audio passthrough at display lost/reset events

thexai opened the pull request:
## Description
Fix audio passthrough at display lost/reset events. 

Fixes https://github.com/xbmc/xbmc/issues/15490


## Motivation and context
I have relied on @FernetMenta's comments to implement this. My systems are not affected by this issue.

> This is a scenario that hasn't been considered in the code. In other words it was working by chance in earlier versions. VideoPlayerAudio checks passthrough capabilities at the time it opens a stream. This can happen at every time, not just at start of playback.
> The mentioned commit introduced no error, it just brought an already existing issue to surface. A audio device change can happen at every time. VideoPlayerAudio needs to re-evaluate passthrough capabilities after a device change and act accordingly.


## How has this been tested?
Tested on RTX 2060 SUPER + Denon AVR-X1600H
and Intel NUC8i3BEK + Denon AVR-X1600H

Although these systems are unaffected and has passthrough audio before and after this PR. 
Kodi log shows the code runs as expected: display reset event triggers re-evaluates passthrough capabilities.

One user affected by this has confirmed that PR fixes the issue:
https://github.com/xbmc/xbmc/issues/15490#issuecomment-838569965

More tests to confirm issue is fixed ~would be desirable...~ two users have already confirmed that it works.

## What is the effect on users?
Fixes missing passthrough audio when Adjust Display Refresh Rate at Start/Stop is enabled.


## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

thexai committed:
Fix audio passthrough at display reset events

a1rwulf commented on xbmc/cores/VideoPlayer/VideoPlayer.cpp:
```diff
@@ -4914,6 +4914,9 @@ void CVideoPlayer::OnResetDisplay()
   m_displayLost = false;
+
+  if (m_VideoPlayerAudio)
+    dynamic_cast<CVideoPlayerAudio*>(m_VideoPlayerAudio)->TriggerCheckPassthrough();
```
Not sure how much I like this cast.
Extending the interface `IDVDStreamPlayerAudio` for this feels weird/wrong as well.
Probably a new `CDVDMsg` would be justified?

thexai replied:
Updated to use the new` CDVDMsg`. In case we choose this option I will squash in a single commit.

a1rwulf replied:
I like it better this way, but maybe @FernetMenta and/or @fritsch can give us their opinion?

fritsch replied:
Mainly the infrastructure is already in place here. The above message / trigger is received cause video player subscribes to the windowing infrastructure. For the dynamic_cast, it could return NULL. If the direct method call is used as interface to establish business logic - add it to the interface, then the cast can be avoided.

The below message comit might be a better idea, but is done asychroneously.

As you see: I don't have a clear opinion. I see the need and understand the code that wants to accomplish that.

thexai replied:
> The below message comit might be a better idea, but is done asychroneously.

I think that in this case, asynchronous, it is also an advantage. If this adds a small delay it is better since it will be re-evaluate passthrough capabilities when the display and audio device are well seated (it could be the case that the display is already available and the audio not yet). With a little delay this possibility is reduced.

FernetMenta replied:
> I like it better this way, but maybe @FernetMenta and/or @fritsch can give us their opinion?

Involving VideoPlayer in the chain of events isn't the clearest solution and does not cover all cases. A change of refresh rate triggers an audio device change event but device change events can occur without a change of refresh rate. The latter case does still not work.
Further re-evaluation of passthrough needs to be done when ActiveAE has completed its reconfiguration after a device change event. Otherwise codec factory might still query outdated capabilities.
PAPlayer needs a solution too.

A proper solution needs to make sure that AE has re-enumerated sinks and adjusted settings: https://github.com/xbmc/xbmc/blob/1fff2cfa6740dae2f0bb501d15558b92fc3befd2/xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp#L629

FernetMenta replied:
Apart from the fact that not all cases are covered this solution is not bad. I agree with @thexai that asynchronous messaging here is much better than the first approach.

thexai replied:
I agree that it does not cover "all" cases but it does cover the case that is causing problems to users: "Audio passthrough does not start initially if Adjust Display Refresh Rate is enabled".

It is preferable to have a "good" solution now than a "perfect" solution that does not exist yet. 😃

thexai replied:
In any case I think the functionality could be extended later if needed:

It is enough to send the message to `m_VideoPlayerAudio` in different ways.

fritsch replied:
Async without clear sync points makes code non deterministic, that's why I stated it. Obviously a message properly sent helps a lot with e.g. concurrency.

thexai commented:
Since we all agree that the second variant is better, I updated the PR with a single commit using `CDVDMsg` method.

Also two uses has confirmed patch is working on Matrix build using the new `CDVDMsg` based code:

https://github.com/xbmc/xbmc/issues/15490#issuecomment-840815403
https://github.com/xbmc/xbmc/issues/15490#issuecomment-840832383

thexai commented:
Updated PR with more elaborated approach. As mentioned @FernetMenta the solution from before does not handle all cases.

Now are used audio device config changed events instead of display reset events. The notification chain has become more complex:

ActiveAE => Application => AppPlayer => VideoPlayer => VideoPlayerAudio

And are used two new async messges:

`TMSG_MEDIA_AUDIO_DEVICE_CHANGED`  (notification from AE to Application)
`PLAYER_AUDIO_DEVICE_CHANGED`  (notification from VideoPlayer to VideoPlayerAudio)

A user has reported that this code works even on a system where the previous solution did not work in all cases:
https://github.com/xbmc/xbmc/issues/15490#issuecomment-842540850

FernetMenta commented:
> ActiveAE => Application => AppPlayer => VideoPlayer => VideoPlayerAudio

Now ist gets ugly :(
A lot of components involved which don't need to know the event and all they do is pass the message on.

FernetMenta commented on xbmc/cores/VideoPlayer/VideoPlayer.cpp:
```diff
@@ -5153,3 +5153,8 @@ void CVideoPlayer::SetUpdateStreamDetails()
+
+void CVideoPlayer::TriggerCheckPassthrough()
+{
+  m_VideoPlayerAudio->SendMessage(new CDVDMsg(CDVDMsg::PLAYER_AUDIO_DEVICE_CHANGED), 1);
```
m_VideoPlayerAudio is not supposed to be accessed fro the outside world.

thexai replied:
VideoPlayer => VideoPlayerAudio   These part has not changed from initial PR....

Only moved from `CVideoPlayer::OnResetDisplay()`  to  `CVideoPlayer::TriggerCheckPassthrough()`

????

FernetMenta replied:
CVideoPlayer::OnResetDisplay() is different. It is a cb VP has registered for. Hence VP has full control when this cb can be called.

thexai commented:
Simplified a bit:
from 12 files changed with 66 additions and 3 deletions.
to
10 files changed with 58 additions and 3 deletions. (removed changes in PlayListPlayer.cpp and Application.h)

Previous commit is here https://github.com/thexai/xbmc/commit/e5d51c36616d3af234276b1f28862e763a0ecad3

I don't think I will spend more time on this PR without a good reason for it (code ugly is not)....

thexai commented:
Reverted to initial and more simple idea of display lost/reset events. This is sufficient to fix 90% of cases and the other 10% is also fixed in combination with "Delay after refresh rate change" setting to 2 - 3 seconds.

thexai commented:
I have been using this PR on a daily basis with no side effects so I think it is safe to merge. In any case, if it is going to be included in v19.2 I think it would be better to merge it as soon as possible so that it can be tested in the nightlies by more people.

@fritsch , do I have your approval to merge?
